File: enrichment.js
import ModelClient, { isUnexpected } from "@azure-rest/ai-inference";
import { AzureKeyCredential } from "@azure/core-auth";
import { Campaign, Event, Session } from './db.js';
import { findBestCampaign } from './campaignMatcher.js';
import { nanoid } from 'nanoid';
import crypto from 'crypto';

let client = null;
const modelName = "openai/gpt-4o-mini";

function getClient() {
  if (!client) {
    const token = process.env["GITHUB_TOKEN"];
    const endpoint = "https://models.github.ai/inference";
    client = ModelClient(endpoint, new AzureKeyCredential(token));
  }
  return client;
}

// --- Performance Layer 1: In-Memory Campaign Cache ---
let campaignCache = [];
let campaignCacheLoaded = false;
let campaignCacheTimestamp = 0;
const CACHE_TTL = 60000;

export async function loadCampaignCache() {
  if (campaignCacheLoaded && Date.now() - campaignCacheTimestamp < CACHE_TTL) {
    return campaignCache;
  }

  const campaigns = await Campaign.find({ active: 1 }).select('+embedding').lean();
  campaignCache = campaigns.map(c => ({
    _id: c._id,
    name: c.name,
    category: c.category,
    brand_url: c.brand_url,
    link_text: c.link_text,
    tracking_code: c.tracking_code,
    description: c.description,
    keywords: c.keywords || [],
    cpc_rate: c.cpc_rate || 0,
    cpa_percentage: c.cpa_percentage || 0,
    tone: c.tone || 'informative',
    embedding: c.embedding,
    keywordSet: new Set((c.keywords || []).map(k => k.toLowerCase())),
  }));
  campaignCacheLoaded = true;
  campaignCacheTimestamp = Date.now();
  console.log(`Campaign cache loaded: ${campaignCache.length} campaigns`);
  return campaignCache;
}

// --- Performance Layer 2: Response Cache ---
const responseCache = new Map();
const RESPONSE_CACHE_TTL = 300000;
const MAX_CACHE_SIZE = 500;

function getCacheKey(prompt, answerOnly = false) {
  const prefix = answerOnly ? 'answer-only-v5:' : 'matcher-v5:';
  return crypto.createHash('md5').update(prefix + prompt.toLowerCase().trim()).digest('hex');
}

function getCachedResponse(prompt, answerOnly = false) {
  const key = getCacheKey(prompt, answerOnly);
  const cached = responseCache.get(key);
  if (cached && Date.now() - cached.timestamp < RESPONSE_CACHE_TTL) {
    return cached.data;
  }
  if (cached) responseCache.delete(key);
  return null;
}

function setCachedResponse(prompt, data, answerOnly = false) {
  if (responseCache.size >= MAX_CACHE_SIZE) {
    const oldestKey = responseCache.keys().next().value;
    responseCache.delete(oldestKey);
  }
  responseCache.set(getCacheKey(prompt, answerOnly), { data, timestamp: Date.now() });
}

// --- Performance Layer 3: Rule-Based Intent Pre-Classification ---
const HOT_KEYWORDS = [
  'sotib', 'olmoq', 'narxi', 'qancha turadi', 'qayerda bor', 'eng arzon',
  'hozir', 'bugun', 'muddatli tolov', "bo'lib to'lash", 'chegirma',
  'aksiya', 'buyurtma', 'yetkazib', 'delivery', 'buy now', 'cheapest',
  'where to buy', 'price', 'deal', 'discount', 'installment', 'купить',
  'цена', 'скидка', 'рассрочка', 'где купить', 'дешевле',
];

const WARM_KEYWORDS = [
  'qaysi biri yaxshi', 'taqqosla', 'farqi', 'review', 'tavsiya',
  'yaxshi', 'yomon', 'compare', 'which is better', 'vs', 'versus',
  'quality', 'warranty', 'какой лучше', 'отзыв', 'сравнить',
  'разница', 'рекомендация',
];

function ruleBasedIntent(prompt) {
  const lower = prompt.toLowerCase();
  let hotScore = 0;
  let warmScore = 0;

  for (const kw of HOT_KEYWORDS) {
    if (lower.includes(kw)) hotScore += 0.3;
  }
  for (const kw of WARM_KEYWORDS) {
    if (lower.includes(kw)) warmScore += 0.3;
  }

  if (hotScore >= 0.6) return { intent: 'hot', confidence: Math.min(hotScore, 1), method: 'rule' };
  if (warmScore >= 0.6) return { intent: 'warm', confidence: Math.min(warmScore, 1), method: 'rule' };
  if (hotScore >= 0.3) return { intent: 'warm', confidence: hotScore, method: 'rule' };

  return null;
}

// --- Main Enrichment Flow ---
export async function enrichResponse(prompt, sessionId = null, options = {}) {
  const { answerOnly = false } = options;
  const startTime = Date.now();
  const session = sessionId || nanoid(16);

  // Check response cache first
  const cached = getCachedResponse(prompt, answerOnly);
  if (cached) {
    console.log(`Response cache HIT (${Date.now() - startTime}ms)`);
    return { ...cached, session_id: session, _cached: true };
  }

  // 1. Intent Classification (Rule-based first, LLM fallback)
  const ruleIntent = ruleBasedIntent(prompt);
  let intentResult;

  if (ruleIntent) {
    intentResult = {
      ...ruleIntent,
      signals: ruleIntent.intent === 'hot'
        ? ['purchase_keywords', 'urgency_detected']
        : ['comparison_keywords'],
      urgency: ruleIntent.intent === 'hot' ? 'high' : 'medium',
      price_sensitivity: 'medium',
      usage: { total_tokens: 0, prompt_tokens: 0, completion_tokens: 0 },
    };
    console.log(`Intent via rules: ${ruleIntent.intent} (${(ruleIntent.confidence * 100).toFixed(0)}%)`);
  } else {
    console.log("Classifying intent via LLM...");
    try {
      intentResult = await classifyIntent(prompt);
    } catch (err) {
      console.error('Intent classification failed:', err.message);
      intentResult = { intent: 'cold', confidence: 0.5, signals: ['llm_failed'], urgency: 'low', price_sensitivity: 'medium', usage: { total_tokens: 0 }, method: 'fallback' };
    }
  }

  // 2–4. Generate response (demo/chat uses answerOnly — never weave brands into the answer body)
  let finalResponse;
  let genUsage = null;
  let appliedCampaign = null;
  let highestScore = -1;
  let matchMethod = 'none';

  if (answerOnly) {
    try {
      const result = await generateStandardContent(prompt, intentResult);
      finalResponse = result.text;
      genUsage = result.usage;
    } catch (err) {
      console.error('Standard response failed:', err.message);
      finalResponse = "Savolingiz bo'yicha yordam bera olaman.";
    }
  } else {
  const campaigns = await loadCampaignCache();
  let bestCampaign = null;

  try {
    const match = await findBestCampaign(prompt, campaigns);
    bestCampaign = match.campaign;
    highestScore = match.score;
    matchMethod = match.method || 'none';
    if (bestCampaign) {
      console.log(`Matched: ${bestCampaign.name} (${match.method}, score: ${highestScore.toFixed(3)}, topic: ${match.dominantTopic || 'n/a'})`);
    } else if (match.dominantTopic) {
      console.log(`No campaign for detected topic "${match.dominantTopic}" — skipping ad`);
    }
  } catch (err) {
    console.error('Campaign matching failed:', err.message);
  }

  if (bestCampaign) {
    appliedCampaign = bestCampaign;
    const generators = {
      hot: generateHotIntentResponse,
      warm: generateWarmIntentResponse,
      cold: generateColdIntentResponse,
    };
    const generator = generators[intentResult.intent] || generateColdIntentResponse;
    console.log(`${intentResult.intent.toUpperCase()} intent → ${bestCampaign.name} (${highestScore.toFixed(3)})`);

    try {
      const result = await generator(prompt, bestCampaign, intentResult);
      finalResponse = result.text;
      genUsage = result.usage;
    } catch (err) {
      console.error('Response generation failed:', err.message);
      finalResponse = `I found something relevant: ${bestCampaign.name} might help. Check them out: ${bestCampaign.brand_url}`;
    }
  } else {
    console.log(`No match (score: ${highestScore > -1 ? highestScore.toFixed(3) : 'N/A'})`);
    try {
      const result = await generateStandardContent(prompt, intentResult);
      finalResponse = result.text;
      genUsage = result.usage;
    } catch (err) {
      console.error('Standard response failed:', err.message);
      finalResponse = "I'm here to help. Could you rephrase your question?";
    }
  }
  }

  // 5. Fire-and-forget analytics
  logAnalytics(session, appliedCampaign, intentResult, highestScore).catch(() => {});

  const result = {
    text: finalResponse,
    session_id: session,
    enriched: !!appliedCampaign,
    campaignId: appliedCampaign ? appliedCampaign._id : null,
    campaign_name: appliedCampaign ? appliedCampaign.name : null,
    category: appliedCampaign ? appliedCampaign.category : 'N/A',
    similarity_score: appliedCampaign ? highestScore.toFixed(3) : null,
    matched_via: appliedCampaign ? matchMethod : null,
    intent: {
      type: intentResult.intent,
      confidence: intentResult.confidence,
      signals: intentResult.signals,
    },
    usage: {
      intent_classification: intentResult.usage,
      generation: genUsage,
      total_tokens: (intentResult.usage?.total_tokens || 0) + (genUsage?.total_tokens || 0),
    },
    performance: {
      total_ms: Date.now() - startTime,
      intent_method: intentResult.method || 'llm',
      match_method: matchMethod,
    },
  };

  if (answerOnly || !appliedCampaign || intentResult.intent === 'cold') {
    setCachedResponse(prompt, result, answerOnly);
  }

  return result;
}

// --- Fire-and-Forget Analytics ---
async function logAnalytics(sessionId, campaign, intent, score) {
  const ops = [];

  if (campaign) {
    ops.push(
      Event.create({ campaign_id: campaign._id, type: 'impression', session_id: sessionId, intent_type: intent.intent }),
      Campaign.findByIdAndUpdate(campaign._id, { $inc: { spent: campaign.cpc_rate, today_impressions: 1 } })
    );
  }

  ops.push(
    Session.create({
      session_id: sessionId,
      user_prompt: sessionId.substring(0, 200),
      intent_type: intent.intent,
      intent_confidence: intent.confidence,
      matched_campaign_id: campaign?._id || null,
      similarity_score: campaign ? score : 0,
      enriched: !!campaign,
    })
  );

  await Promise.allSettled(ops);
}

// --- LLM Intent Classification (Fallback) ---
async function classifyIntent(prompt) {
  const response = await getClient().path("/chat/completions").post({
    body: {
      messages: [
        {
          role: "system",
          content: `Classify intent: COLD (exploring), WARM (comparing), HOT (buying now).
JSON: {"intent":"cold|warm|hot","confidence":0.0-1.0,"signals":["s1"],"urgency":"low|medium|high","price_sensitivity":"low|medium|high"}`
        },
        { role: "user", content: prompt }
      ],
      model: modelName,
      temperature: 0.1,
      max_tokens: 150,
    }
  });

  if (isUnexpected(response)) throw response.body.error;

  let result;
  try {
    const content = response.body.choices[0].message.content;
    const jsonMatch = content.match(/\{[\s\S]*\}/);
    result = JSON.parse(jsonMatch[0]);
  } catch {
    result = { intent: 'cold', confidence: 0.5, signals: ['parse_failed'], urgency: 'low', price_sensitivity: 'medium' };
  }

  result.usage = response.body.usage;
  result.method = 'llm';
  return result;
}

// --- Response Generators ---
async function generateHotIntentResponse(prompt, campaign, intent) {
  const systemPrompt = `Synaptic AI — Intent-Based Conversational Commerce.
INTENT: HOT (buy now) | URGENCY: ${intent.urgency}
RULES: 1) Direct answer first 2) Natural recommendation 3) Mention deals/muddatli to'lov 4) NEVER paste URLs — only mention brand by name: ${campaign.link_text || campaign.name} 5) Concise 6) Match language.
BRAND: ${campaign.name} | ${campaign.category} | ${campaign.description}`;

  const response = await getClient().path("/chat/completions").post({
    body: { messages: [{ role: "system", content: systemPrompt }, { role: "user", content: prompt }], model: modelName, temperature: 0.3 }
  });
  if (isUnexpected(response)) throw response.body.error;
  return { text: response.body.choices[0].message.content, usage: response.body.usage };
}

async function generateWarmIntentResponse(prompt, campaign, intent) {
  const systemPrompt = `Synaptic AI — Intent-Based Conversational Commerce.
INTENT: WARM (comparing) | URGENCY: ${intent.urgency}
RULES: 1) Comparative info 2) Educational 3) Campaign as one option 4) NEVER paste URLs — brand name only: ${campaign.link_text || campaign.name} 5) Match language.
BRAND: ${campaign.name} | ${campaign.category} | ${campaign.description}`;

  const response = await getClient().path("/chat/completions").post({
    body: { messages: [{ role: "system", content: systemPrompt }, { role: "user", content: prompt }], model: modelName, temperature: 0.4 }
  });
  if (isUnexpected(response)) throw response.body.error;
  return { text: response.body.choices[0].message.content, usage: response.body.usage };
}

async function generateColdIntentResponse(prompt, campaign, intent) {
  const systemPrompt = `Synaptic AI — Intent-Based Conversational Commerce.
INTENT: COLD (exploring) | TOPIC: ${campaign.category}
RULES: 1) Answer the user's question fully and accurately first 2) Only mention the brand at the end IF it directly relates to what the user asked 3) Never suggest unrelated products 4) NEVER paste URLs — brand name only: ${campaign.link_text || campaign.name} 5) Match language.
BRAND (use only if relevant): ${campaign.name} | ${campaign.description}`;

  const response = await getClient().path("/chat/completions").post({
    body: { messages: [{ role: "system", content: systemPrompt }, { role: "user", content: prompt }], model: modelName, temperature: 0.5 }
  });
  if (isUnexpected(response)) throw response.body.error;
  return { text: response.body.choices[0].message.content, usage: response.body.usage };
}

async function generateStandardContent(prompt, intent) {
  const response = await getClient().path("/chat/completions").post({
    body: {
      messages: [
        {
          role: "system",
          content: `Synaptic AI assistant. Intent: ${intent.intent.toUpperCase()}.
Answer the user's question directly and completely. Match their language (Uzbek/Russian/English).
Do NOT mention sponsors, ads, brands, stores, or product links.
For informational questions (sports, companies, universities, visas, statistics) give facts only — never pivot to unrelated shopping.`,
        },
        { role: "user", content: prompt },
      ],
      model: modelName,
    }
  });
  if (isUnexpected(response)) throw response.body.error;
  return { text: response.body.choices[0].message.content, usage: response.body.usage };
}

File: backend/app/services/enrichment.py
import json
import logging
import re
import time
from functools import lru_cache

from nanoid import generate as nanoid
from openai import AsyncOpenAI

from app.config import settings
from app.core.bson_utils import oid
from app.db.mongo import get_db
from app.services.campaign_cache import load_campaign_cache
from app.services.campaign_cache import increment_campaign_impression
from app.services.conversation import MARKDOWN_SYSTEM_HINT, build_matching_prompt, history_cache_key, normalize_chat_history
from app.services.matcher import find_best_campaign

logger = logging.getLogger(__name__)

MODEL = "openai/gpt-4o-mini"
HOT_KEYWORDS = [
    "sotib", "olmoq", "narxi", "qancha turadi", "qayerda bor", "eng arzon", "hozir", "bugun",
    "muddatli tolov", "bo'lib to'lash", "chegirma", "aksiya", "buyurtma", "yetkazib", "delivery",
    "buy now", "cheapest", "where to buy", "price", "deal", "discount", "installment",
]
WARM_KEYWORDS = [
    "qaysi biri yaxshi", "taqqosla", "farqi", "review", "tavsiya", "yaxshi", "yomon",
    "compare", "which is better", "vs", "versus", "quality", "warranty",
]

_response_cache: dict = {}
RESPONSE_TTL = 300_000
MAX_CACHE = 500


@lru_cache
def _client() -> AsyncOpenAI | None:
    if not settings.github_token:
        return None
    return AsyncOpenAI(base_url="https://models.github.ai/inference", api_key=settings.github_token)


def _rule_intent(prompt: str) -> dict | None:
    lower = prompt.lower()
    hot = sum(0.3 for kw in HOT_KEYWORDS if kw in lower)
    warm = sum(0.3 for kw in WARM_KEYWORDS if kw in lower)
    if hot >= 0.6:
        return {"intent": "hot", "confidence": min(hot, 1.0), "method": "rule"}
    if warm >= 0.6:
        return {"intent": "warm", "confidence": min(warm, 1.0), "method": "rule"}
    if hot >= 0.3:
        return {"intent": "warm", "confidence": hot, "method": "rule"}
    return None


def _get_cached(prompt, history, answer_only):
    key = history_cache_key(prompt, history, answer_only)
    cached = _response_cache.get(key)
    if cached and time.time() * 1000 - cached["ts"] < RESPONSE_TTL:
        return cached["data"]
    if cached:
        _response_cache.pop(key, None)
    return None


def _set_cached(prompt, history, data, answer_only):
    if len(_response_cache) >= MAX_CACHE:
        _response_cache.pop(next(iter(_response_cache)))
    _response_cache[history_cache_key(prompt, history, answer_only)] = {"data": data, "ts": time.time() * 1000}


def _build_llm_messages(system_prompt, history, latest, use_markdown):
    system = f"{system_prompt}\n\n{MARKDOWN_SYSTEM_HINT}" if use_markdown else system_prompt
    msgs = [{"role": "system", "content": system}]
    turns = normalize_chat_history(history)
    prompt = str(latest or "").strip()
    last_is_current = turns and turns[-1]["role"] == "user" and turns[-1]["content"] == prompt
    prior = turns[:-1] if last_is_current else turns
    msgs.extend({"role": t["role"], "content": t["content"]} for t in prior)
    msgs.append({"role": "user", "content": prompt})
    return msgs


async def _chat(system_prompt, history, latest, use_markdown=True, temperature=0.4):
    client = _client()
    if not client:
        raise RuntimeError("GITHUB_TOKEN not set")
    resp = await client.chat.completions.create(
        model=MODEL,
        messages=_build_llm_messages(system_prompt, history, latest, use_markdown),
        temperature=temperature,
    )
    return {"text": resp.choices[0].message.content or "", "usage": resp.usage.model_dump() if resp.usage else {}}


async def _classify_intent(prompt: str) -> dict:
    client = _client()
    resp = await client.chat.completions.create(
        model=MODEL,
        messages=[
            {
                "role": "system",
                "content": 'Classify intent: COLD (exploring), WARM (comparing), HOT (buying now).\nJSON: {"intent":"cold|warm|hot","confidence":0.0-1.0,"signals":["s1"],"urgency":"low|medium|high","price_sensitivity":"low|medium|high"}',
            },
            {"role": "user", "content": prompt},
        ],
        temperature=0.1,
        max_tokens=150,
    )
    content = resp.choices[0].message.content or ""
    try:
        m = re.search(r"\{[\s\S]*\}", content)
        result = json.loads(m.group(0)) if m else {"intent": "cold", "confidence": 0.5, "signals": ["parse_failed"]}
    except json.JSONDecodeError:
        result = {"intent": "cold", "confidence": 0.5, "signals": ["parse_failed"]}
    result["usage"] = resp.usage.model_dump() if resp.usage else {}
    result["method"] = "llm"
    result.setdefault("urgency", "low")
    result.setdefault("price_sensitivity", "medium")
    return result


async def _log_analytics(session_id, campaign, intent, score):
    db = get_db()
    ops = []
    if campaign:
        await db.events.insert_one(
            {
                "campaign_id": oid(campaign["_id"]),
                "type": "impression",
                "session_id": session_id,
                "intent_type": intent.get("intent"),
                "createdAt": __import__("datetime").datetime.now(__import__("datetime").timezone.utc),
            }
        )
        await increment_campaign_impression(str(campaign["_id"]), campaign.get("cpc_rate") or 0)
    await db.sessions.insert_one(
        {
            "session_id": session_id,
            "user_prompt": session_id[:200],
            "intent_type": intent.get("intent", "cold"),
            "intent_confidence": intent.get("confidence", 0),
            "matched_campaign_id": oid(campaign["_id"]) if campaign else None,
            "similarity_score": score if campaign else 0,
            "enriched": bool(campaign),
            "createdAt": __import__("datetime").datetime.now(__import__("datetime").timezone.utc),
        }
    )


async def enrich_response(prompt: str, session_id: str | None = None, *, answer_only=False, messages=None, parse_mode="Markdown"):
    start = time.time() * 1000
    session = session_id or nanoid(size=16)
    history = normalize_chat_history(messages or [])
    match_prompt = build_matching_prompt(prompt, history)
    use_md = parse_mode.lower() == "markdown"

    cached = _get_cached(prompt, history, answer_only)
    if cached and len(history) <= 2:
        return {**cached, "session_id": session, "_cached": True}

    rule = _rule_intent(match_prompt)
    if rule:
        intent = {
            **rule,
            "signals": ["purchase_keywords", "urgency_detected"] if rule["intent"] == "hot" else ["comparison_keywords"],
            "urgency": "high" if rule["intent"] == "hot" else "medium",
            "price_sensitivity": "medium",
            "usage": {"total_tokens": 0, "prompt_tokens": 0, "completion_tokens": 0},
        }
    else:
        try:
            intent = await _classify_intent(match_prompt)
        except Exception as err:
            logger.warning("Intent classification failed: %s", err)
            intent = {"intent": "cold", "confidence": 0.5, "signals": ["llm_failed"], "urgency": "low", "price_sensitivity": "medium", "usage": {}, "method": "fallback"}

    applied = None
    score = -1.0
    match_method = "none"
    gen_usage = None
    final_text = ""

    if answer_only:
        try:
            r = await _chat(
                f"Synaptic AI assistant. Intent: {intent['intent'].upper()}.\nAnswer directly. No ads/brands/URLs.",
                history,
                prompt,
                use_md,
            )
            final_text = r["text"]
            gen_usage = r["usage"]
        except Exception:
            final_text = "Savolingiz bo'yicha yordam bera olaman."
    else:
        campaigns = await load_campaign_cache()
        try:
            match = await find_best_campaign(
                prompt,
                campaigns,
                serving_context={"sessionId": session},
                match_prompt=match_prompt,
            )
            applied = match.get("campaign")
            score = match.get("score", -1)
            match_method = match.get("method") or "none"
        except Exception as err:
            logger.warning("Campaign matching failed: %s", err)

        if applied:
            brand = applied.get("link_text") or applied.get("name")
            templates = {
                "hot": f"Synaptic AI. HOT intent. Mention brand by name only: {brand}. BRAND: {applied.get('name')} | {applied.get('description')}",
                "warm": f"Synaptic AI. WARM intent. Comparative tone. Brand: {brand}.",
                "cold": f"Synaptic AI. COLD intent. Topic: {applied.get('category')}. Brand if relevant: {brand}.",
            }
            sys_p = templates.get(intent.get("intent"), templates["cold"])
            try:
                r = await _chat(sys_p, history, prompt, use_md, 0.35 if intent.get("intent") == "hot" else 0.45)
                final_text = r["text"]
                gen_usage = r["usage"]
            except Exception:
                final_text = f"I found something relevant: {applied.get('name')} might help."
        else:
            try:
                r = await _chat(
                    f"Synaptic AI. Intent: {intent.get('intent', 'cold').upper()}. Answer fully. No ads.",
                    history,
                    prompt,
                    use_md,
                )
                final_text = r["text"]
                gen_usage = r["usage"]
            except Exception:
                final_text = "I'm here to help. Could you rephrase your question?"

    try:
        await _log_analytics(session, applied, intent, score)
    except Exception:
        pass

    result = {
        "text": final_text,
        "session_id": session,
        "enriched": bool(applied),
        "campaignId": applied["_id"] if applied else None,
        "campaign_name": applied.get("name") if applied else None,
        "category": applied.get("category") if applied else "N/A",
        "similarity_score": f"{score:.3f}" if applied and score > -1 else None,
        "matched_via": match_method if applied else None,
        "intent": {"type": intent.get("intent"), "confidence": intent.get("confidence"), "signals": intent.get("signals", [])},
        "usage": {
            "intent_classification": intent.get("usage"),
            "generation": gen_usage,
            "total_tokens": (intent.get("usage") or {}).get("total_tokens", 0) + (gen_usage or {}).get("total_tokens", 0),
        },
        "performance": {
            "total_ms": int(time.time() * 1000 - start),
            "intent_method": intent.get("method", "llm"),
            "match_method": match_method,
        },
    }

    if (answer_only or not applied or intent.get("intent") == "cold") and len(history) <= 2:
        _set_cached(prompt, history, result, answer_only)

    return result
